neuralpp/symbolic/expression.py

```diff
@@ -471,6 +479,51 @@ def dict(self) -> Dict[str, Any]:
         pass
 
 
+class QuantifierExpression(Expression, ABC):
+    @property
+    @abstractmethod
+    def operation(self) -> Constant:
+        """ operation is a Constant wrapping a function.
+        E.g., integer Summation's operation is Constant(operator.add, Callable[[int,int],int]). """
+        pass
+
+    @property
+    @abstractmethod
+    def index(self) -> Variable:
+        pass
+
+    @property
+    @abstractmethod
+    def constrain(self) -> Expression:
+        """ User can expect the returning `expression` to be a Boolean Expression, i.e., expression.type == bool. """
+        pass
+
+    @property
+    @abstractmethod
+    def body(self) -> Expression:
+        pass
+
+    @property
+    def subexpressions(self) -> List[Expression]:
+        return [self.operation, self.index, self.constrain, self.body]
+
+    def set(self, i: int, new_expression: Expression) -> Expression:
+        subexpressions = self.subexpressions
+        subexpressions[i] = new_expression
+        return self.new_quantifier_expression(*subexpressions)
+
+    def replace(self, from_expression: Expression, to_expression: Expression) -> Expression:
+        if from_expression.syntactic_eq(self):
+            return to_expression
+
+        # recursively do the replacement
+        new_subexpressions = [
+            to_expression if e.syntactic_eq(from_expression) else e.replace(from_expression, to_expression)
```

Comment:
We might not even need to check `e.syntactic_eq` here because `e.replace` cover that case already. (ditto to the definition of `FunctionApplication.replace`)